include/envoy/http/filter.h

```diff
@@ -240,6 +240,12 @@ class StreamDecoderFilterCallbacks : public virtual StreamFilterCallbacks {
    */
   virtual void modifyDecodingBuffer(std::function<void(Buffer::Instance&)> callback) PURE;
 
+  /**
+   * Allows modifying the decoding header. May only be called before any header has been continued
+   * past the calling filter.
+   */
+  virtual void modifyDecodingHeaders(std::function<void(Http::HeaderMap&)> callback) PURE;
```

Comment:
These functions aren't actually needed. You can store and modify the headers when you get decode/encode headers.

Reply:
Hmm, maybe I got it wrong. But, from what I experienced, a filter has an access to the request/response header map only when encode/decodeHeader is invoked. If the filter wants to append custom header after inspecting the payload (en,decodeData function), it doesn't have a way to modify the header, unless modifyDecodingHeaders function can handle? Please feel free to correct me if I got it wrong.

Reply:
You can store a reference to the header map and modify it later before you continue iteration.